{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b391df05",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/jessica/mambaforge/envs/scanpy/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "import scanpy.external as sce\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import scvi\n",
    "import torch\n",
    "import scipy.sparse as sp\n",
    "\n",
    "from rich import print\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import os\n",
    "\n",
    "outdir = \"/Users/jessica/Documents/GitHub/Jessica_notebooks/write/02_batch_annotation/PICA0001-PICA0007\"\n",
    "os.makedirs(outdir, exist_ok=True)\n",
    "sc.settings.figdir = \"/Users/jessica/Documents/GitHub/Jessica_notebooks/figures/PICA0001-PICA0007/02_annotation\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "87253030",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">AnnData object with n_obs × n_vars = <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">10059</span> × <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">38606</span>\n",
       "    obs: <span style=\"color: #008000; text-decoration-color: #008000\">'status'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'assignment'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'pica_id'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'pool_id'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'sequencing_batch'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'n_genes_by_counts'</span>, \n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">'log1p_n_genes_by_counts'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'total_counts'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'log1p_total_counts'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'pct_counts_in_top_50_genes'</span>, \n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">'pct_counts_in_top_100_genes'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'pct_counts_in_top_200_genes'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'pct_counts_in_top_500_genes'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'total_counts_mt'</span>, \n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">'log1p_total_counts_mt'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'pct_counts_mt'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'total_counts_ribo'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'log1p_total_counts_ribo'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'pct_counts_ribo'</span>, \n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">'total_counts_hb'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'log1p_total_counts_hb'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'pct_counts_hb'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'status_manual'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'cell_type'</span>, \n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">'status_manual_01_b_subset'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'broad_cell_type'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'cell_type_01_b_cell'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'cell_type_01_b_cell_corrected'</span>, \n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">'cell_type_02_t_cell_broad'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'cell_type_03_t_cell_subtype'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'cell_type_04_myeloid_subset'</span>\n",
       "    var: <span style=\"color: #008000; text-decoration-color: #008000\">'gene_ids'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'feature_types'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'genome'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'mt'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'ribo'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'hb'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'n_cells_by_counts-0'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'mean_counts-0'</span>, \n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">'log1p_mean_counts-0'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'pct_dropout_by_counts-0'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'total_counts-0'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'log1p_total_counts-0'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'n_cells_by_counts-1'</span>, \n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">'mean_counts-1'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'log1p_mean_counts-1'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'pct_dropout_by_counts-1'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'total_counts-1'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'log1p_total_counts-1'</span>, \n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">'n_cells_by_counts-2'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'mean_counts-2'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'log1p_mean_counts-2'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'pct_dropout_by_counts-2'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'total_counts-2'</span>, \n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">'log1p_total_counts-2'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'highly_variable'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'means'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'dispersions'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'dispersions_norm'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'highly_variable_nbatches'</span>, \n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">'highly_variable_intersection'</span>\n",
       "    uns: <span style=\"color: #008000; text-decoration-color: #008000\">'hvg'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'log1p'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'neighbors'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'pca'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'pica_id_colors'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'umap'</span>\n",
       "    obsm: <span style=\"color: #008000; text-decoration-color: #008000\">'X_pca'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'X_umap'</span>\n",
       "    varm: <span style=\"color: #008000; text-decoration-color: #008000\">'PCs'</span>\n",
       "    layers: <span style=\"color: #008000; text-decoration-color: #008000\">'counts'</span>\n",
       "    obsp: <span style=\"color: #008000; text-decoration-color: #008000\">'connectivities'</span>, <span style=\"color: #008000; text-decoration-color: #008000\">'distances'</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "AnnData object with n_obs × n_vars = \u001b[1;36m10059\u001b[0m × \u001b[1;36m38606\u001b[0m\n",
       "    obs: \u001b[32m'status'\u001b[0m, \u001b[32m'assignment'\u001b[0m, \u001b[32m'pica_id'\u001b[0m, \u001b[32m'pool_id'\u001b[0m, \u001b[32m'sequencing_batch'\u001b[0m, \u001b[32m'n_genes_by_counts'\u001b[0m, \n",
       "\u001b[32m'log1p_n_genes_by_counts'\u001b[0m, \u001b[32m'total_counts'\u001b[0m, \u001b[32m'log1p_total_counts'\u001b[0m, \u001b[32m'pct_counts_in_top_50_genes'\u001b[0m, \n",
       "\u001b[32m'pct_counts_in_top_100_genes'\u001b[0m, \u001b[32m'pct_counts_in_top_200_genes'\u001b[0m, \u001b[32m'pct_counts_in_top_500_genes'\u001b[0m, \u001b[32m'total_counts_mt'\u001b[0m, \n",
       "\u001b[32m'log1p_total_counts_mt'\u001b[0m, \u001b[32m'pct_counts_mt'\u001b[0m, \u001b[32m'total_counts_ribo'\u001b[0m, \u001b[32m'log1p_total_counts_ribo'\u001b[0m, \u001b[32m'pct_counts_ribo'\u001b[0m, \n",
       "\u001b[32m'total_counts_hb'\u001b[0m, \u001b[32m'log1p_total_counts_hb'\u001b[0m, \u001b[32m'pct_counts_hb'\u001b[0m, \u001b[32m'status_manual'\u001b[0m, \u001b[32m'cell_type'\u001b[0m, \n",
       "\u001b[32m'status_manual_01_b_subset'\u001b[0m, \u001b[32m'broad_cell_type'\u001b[0m, \u001b[32m'cell_type_01_b_cell'\u001b[0m, \u001b[32m'cell_type_01_b_cell_corrected'\u001b[0m, \n",
       "\u001b[32m'cell_type_02_t_cell_broad'\u001b[0m, \u001b[32m'cell_type_03_t_cell_subtype'\u001b[0m, \u001b[32m'cell_type_04_myeloid_subset'\u001b[0m\n",
       "    var: \u001b[32m'gene_ids'\u001b[0m, \u001b[32m'feature_types'\u001b[0m, \u001b[32m'genome'\u001b[0m, \u001b[32m'mt'\u001b[0m, \u001b[32m'ribo'\u001b[0m, \u001b[32m'hb'\u001b[0m, \u001b[32m'n_cells_by_counts-0'\u001b[0m, \u001b[32m'mean_counts-0'\u001b[0m, \n",
       "\u001b[32m'log1p_mean_counts-0'\u001b[0m, \u001b[32m'pct_dropout_by_counts-0'\u001b[0m, \u001b[32m'total_counts-0'\u001b[0m, \u001b[32m'log1p_total_counts-0'\u001b[0m, \u001b[32m'n_cells_by_counts-1'\u001b[0m, \n",
       "\u001b[32m'mean_counts-1'\u001b[0m, \u001b[32m'log1p_mean_counts-1'\u001b[0m, \u001b[32m'pct_dropout_by_counts-1'\u001b[0m, \u001b[32m'total_counts-1'\u001b[0m, \u001b[32m'log1p_total_counts-1'\u001b[0m, \n",
       "\u001b[32m'n_cells_by_counts-2'\u001b[0m, \u001b[32m'mean_counts-2'\u001b[0m, \u001b[32m'log1p_mean_counts-2'\u001b[0m, \u001b[32m'pct_dropout_by_counts-2'\u001b[0m, \u001b[32m'total_counts-2'\u001b[0m, \n",
       "\u001b[32m'log1p_total_counts-2'\u001b[0m, \u001b[32m'highly_variable'\u001b[0m, \u001b[32m'means'\u001b[0m, \u001b[32m'dispersions'\u001b[0m, \u001b[32m'dispersions_norm'\u001b[0m, \u001b[32m'highly_variable_nbatches'\u001b[0m, \n",
       "\u001b[32m'highly_variable_intersection'\u001b[0m\n",
       "    uns: \u001b[32m'hvg'\u001b[0m, \u001b[32m'log1p'\u001b[0m, \u001b[32m'neighbors'\u001b[0m, \u001b[32m'pca'\u001b[0m, \u001b[32m'pica_id_colors'\u001b[0m, \u001b[32m'umap'\u001b[0m\n",
       "    obsm: \u001b[32m'X_pca'\u001b[0m, \u001b[32m'X_umap'\u001b[0m\n",
       "    varm: \u001b[32m'PCs'\u001b[0m\n",
       "    layers: \u001b[32m'counts'\u001b[0m\n",
       "    obsp: \u001b[32m'connectivities'\u001b[0m, \u001b[32m'distances'\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "raw_adata_cd4= sc.read_h5ad(\"/Users/jessica/Documents/GitHub/Jessica_notebooks/write/02_batch_annotation/PICA0001-PICA0007/PICA0001-PICA0007_preprocessed.h5ad\")\n",
    "raw_adata_cd4.var_names_make_unique()\n",
    "print(raw_adata_cd4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1822a731",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "cell_type",
         "rawType": "category",
         "type": "unknown"
        },
        {
         "name": "count",
         "rawType": "int64",
         "type": "integer"
        }
       ],
       "ref": "ae8e00aa-0ec2-4d9d-89a5-c0f4ce2b9e65",
       "rows": [
        [
         "Naive/Central memory CD4 T cell",
         "9527"
        ],
        [
         "Regulatory T cell",
         "267"
        ],
        [
         "CCR6+ memory CD4 T cell",
         "222"
        ],
        [
         "Cytotoxic CD4 T cell",
         "43"
        ]
       ],
       "shape": {
        "columns": 1,
        "rows": 4
       }
      },
      "text/plain": [
       "cell_type\n",
       "Naive/Central memory CD4 T cell    9527\n",
       "Regulatory T cell                   267\n",
       "CCR6+ memory CD4 T cell             222\n",
       "Cytotoxic CD4 T cell                 43\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_adata_cd4.obs['cell_type'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2800a4e2",
   "metadata": {},
   "source": [
    "#### 1. Batch integration with scVI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d055a271",
   "metadata": {},
   "outputs": [],
   "source": [
    "scvi.model.SCVI.setup_anndata(raw_adata_cd4, layer=\"counts\", batch_key=\"pica_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8955594a",
   "metadata": {},
   "outputs": [],
   "source": [
    "model = scvi.model.SCVI(raw_adata_cd4, n_layers=2, n_latent=30, gene_likelihood=\"nb\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5482b8d0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "GPU available: True (mps), used: False\n",
      "TPU available: False, using: 0 TPU cores\n",
      "HPU available: False, using: 0 HPUs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 26/400:   6%|▋         | 25/400 [09:23<3:15:36, 31.30s/it, v_num=1, train_loss_step=8.32e+3, train_loss_epoch=8.15e+3]"
     ]
    }
   ],
   "source": [
    "model.train()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6aac0e6f",
   "metadata": {},
   "outputs": [
    {
     "ename": "RuntimeError",
     "evalue": "Trying to query inferred values from an untrained model. Please train the model first.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[6], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m raw_adata_cd4\u001b[38;5;241m.\u001b[39mobsm[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mX_scVI\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[43mmodel\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_latent_representation\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;66;03m#  latent representation \u001b[39;00m\n\u001b[1;32m      4\u001b[0m sc\u001b[38;5;241m.\u001b[39mpp\u001b[38;5;241m.\u001b[39mneighbors(raw_adata_cd4, use_rep\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mX_scVI\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m~/mambaforge/envs/scanpy/lib/python3.10/site-packages/torch/utils/_contextlib.py:120\u001b[0m, in \u001b[0;36mcontext_decorator.<locals>.decorate_context\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    117\u001b[0m \u001b[38;5;129m@functools\u001b[39m\u001b[38;5;241m.\u001b[39mwraps(func)\n\u001b[1;32m    118\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21mdecorate_context\u001b[39m(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m    119\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m ctx_factory():\n\u001b[0;32m--> 120\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/mambaforge/envs/scanpy/lib/python3.10/site-packages/scvi/model/base/_vaemixin.py:300\u001b[0m, in \u001b[0;36mVAEMixin.get_latent_representation\u001b[0;34m(self, adata, indices, give_mean, mc_samples, batch_size, return_dist, dataloader)\u001b[0m\n\u001b[1;32m    296\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mtorch\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mnn\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mfunctional\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m softmax\n\u001b[1;32m    298\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mscvi\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mmodule\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01m_constants\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m MODULE_KEYS\n\u001b[0;32m--> 300\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_check_if_trained\u001b[49m\u001b[43m(\u001b[49m\u001b[43mwarn\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[1;32m    301\u001b[0m _validate_adata_dataloader_input(\u001b[38;5;28mself\u001b[39m, adata, dataloader)\n\u001b[1;32m    303\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m dataloader \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/mambaforge/envs/scanpy/lib/python3.10/site-packages/scvi/model/base/_base_model.py:559\u001b[0m, in \u001b[0;36mBaseModelClass._check_if_trained\u001b[0;34m(self, warn, message)\u001b[0m\n\u001b[1;32m    557\u001b[0m     warnings\u001b[38;5;241m.\u001b[39mwarn(message, \u001b[38;5;167;01mUserWarning\u001b[39;00m, stacklevel\u001b[38;5;241m=\u001b[39msettings\u001b[38;5;241m.\u001b[39mwarnings_stacklevel)\n\u001b[1;32m    558\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 559\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m(message)\n",
      "\u001b[0;31mRuntimeError\u001b[0m: Trying to query inferred values from an untrained model. Please train the model first."
     ]
    }
   ],
   "source": [
    "raw_adata_cd4.obsm[\"X_scVI\"] = model.get_latent_representation()\n",
    "\n",
    "#  latent representation \n",
    "sc.pp.neighbors(raw_adata_cd4, use_rep=\"X_scVI\")\n",
    "sc.tl.umap(raw_adata_cd4)\n",
    "\n",
    "\n",
    "sc.pl.umap(raw_adata_cd4, color=[ \"pica_id\", 'cell_type'], save=\"_scvi_integrated_umap.png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f24ced2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(\n",
    "    raw_adata_cd4,\n",
    "    color=[\"log1p_total_counts\", \"pct_counts_mt\", \"log1p_n_genes_by_counts\"],\n",
    "    wspace=0.5,\n",
    "    ncols=2,\n",
    "    save=\"_scvi_qc.png\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fcd14b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the scVI integrated batch adata\n",
    "raw_adata_cd4.write_h5ad(\"/Users/jessica/Documents/GitHub/Jessica_notebooks/write/02_batch_annotation//PICA0001-PICA0007/PICA0001-PICA0007_scVI.h5ad\", compression='gzip')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e69e224b",
   "metadata": {},
   "source": [
    "#### 2. Clustering\n",
    "- resolutions\n",
    "- known marker genes to see if they align with cluster size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b549944b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Test different resolutions for Leiden clustering\n",
    "# Higher resolution will result in more clusters, lower resolution will result in fewer clusters\n",
    "for res in [ 0.5, 1.0, 1.5, 2.0]:\n",
    "    sc.tl.leiden(\n",
    "        raw_adata_cd4, key_added=f\"leiden_res_{res:3.1f}\", resolution=res, flavor=\"igraph\"\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "244f16e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(\n",
    "    raw_adata_cd4,\n",
    "    color=[ \"leiden_res_0.5\", \"leiden_res_1.0\",\"leiden_res_1.5\", \"leiden_res_2.0\"],\n",
    "    legend_loc=\"on data\", legend_fontsize = 15\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9784512",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the clustered adata\n",
    "raw_adata_cd4.write_h5ad(\"/Users/jessica/Documents/GitHub/Jessica_notebooks/write/02_batch_annotation/PICA0001-PICA0007/PICA0001-PICA0007_scVI_cluster.h5ad\", compression='gzip')\n",
    "print(raw_adata_cd4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79e33ef3",
   "metadata": {},
   "outputs": [],
   "source": [
    "raw_adata_cd4= sc.read_h5ad(\"/Users/jessica/Documents/GitHub/Jessica_notebooks/write/02_batch_annotation/PICA0001-PICA0007/PICA0001-PICA0007_scVI_cluster.h5ad\")\n",
    "print(raw_adata_cd4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4eeaa438",
   "metadata": {},
   "source": [
    "##### 3. CD4 subset validation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cf430ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "broad_marker_genes = {\n",
    "    \"B cell\": [\"BLNK\", \"CD19\", \"CD79A\", \"CD79B\", \"MS4A1\", \"CD37\", \"CD79A\"],\n",
    "    \"Plasmablast\": [\"CD19\", \"CD24\", \"CD38\"], # experiment\n",
    "    \"Dendritic cell\": [\"CLEC4C\", \"IL3RA\", \"NRP1\", \"CD1C\", \"CST3\", \"FCER1A\", \"CD123\", \"GZMB\"],\n",
    "    \"Endothelial cell\": [\"CDH5\", \"SELE\", \"VWF\"],\n",
    "    \"Fibroblast\": [\"COL1A1\", \"COL3A1\", \"FAP\", \"THY1\"],\n",
    "    'Neutrophils': ['S100A8', 'S100A9', 'CEACAM8', 'ELANE', 'CSF3R', 'MPO'],  # CEACAM8 is highly specific neutrophil marker\n",
    "    'Eosinophils': ['IL5RA', 'CCR3', 'EPX', 'PRG2', 'GATA1', 'CLC'],  \n",
    "    'Basophils': ['FCER1A', 'KIT', 'CPA3', 'HPGDS', 'ENPP3'],  \n",
    "    \"Classical monocyte\": [\"CD14\", \"FCGR3A\", \"LYZ\", \"FCGR1A\", 'CCR2', 'S100A8', 'S100A9'], # CD14⁺ CD16/FCGR3A⁻\n",
    "    \"Intermediate monocyte\": [\"CD14\", \"FCGR3A\", \"HLA-DRA\", \"CCL2\", 'ITGAX'], # CD14⁺ CD16⁺\n",
    "    \"Non-classical monocyte\": [\"CD14\", \"FCGR3A\", \"TREML4\", \"NR4A1\", 'CX3CR1', 'HLA-DPB1'], # CD14⁻CD16⁺\n",
    "    \"Natural killer cell\": [\"CCL3\", 'CD160', 'CD247', 'GNLY', 'GZMB', 'NKG7', 'FCGR3A', 'FCGR3B', 'KLRB1', 'KLRC1', 'KLRD1', 'KLRF1', 'KLRK1', 'NCAM1'],\n",
    "    \"Platelet\": [\"CD41\", \"ITGA2B\", \"CD34\", \"CD61\", \"PF4\", 'PLA2G12A', 'PPBP'], # No megakaryocyte in blood\n",
    "    'HSPC': ['CD34', 'CD38', 'THY1', 'KIT', 'PROM1', 'SLAMF1', 'ITGA6'],\n",
    "    \"T cell\": ['CD3D', 'CD3E', 'CD3G', 'TRAC', 'TRBC1', 'TRDC'],\n",
    "    \"CD4+ memory T cell\": ['CCR7', 'CD27', 'IL7R'],\n",
    "    \"CD4+ T cell\": ['CD4', 'CTLA4', 'FOXP3', 'IL2RA'],\n",
    "    \"CD8+ T cell\": ['CD4', 'CD8A', 'CD8B', 'GZMK'],\n",
    "    \"Naive T cell\": ['CCR7', 'CD27', 'CD8A', 'CD8B'],\n",
    "    \"Natural killer T (NKT) cell\": ['CD8A', 'CD8B', 'ZNF683'],\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3f0f0dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_full = raw_adata_cd4.raw.to_adata()\n",
    "adata_full\n",
    "present_marker_genes = {}\n",
    "\n",
    "for celltype, markers in broad_marker_genes.items():\n",
    "    present_markers = []\n",
    "    for marker in markers:\n",
    "        if marker in adata_full.var_names:\n",
    "            present_markers.append(marker)\n",
    "    present_marker_genes[celltype] = present_markers\n",
    "print(present_marker_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e350749",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.dotplot(raw_adata_cd4, present_marker_genes , groupby=\"leiden_res_1.0\", standard_scale=\"var\", save=\"_all_markers.png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9465c600",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.dotplot(raw_adata_cd4, [\"CD19\", \"MS4A1\", \"CD79A\", \"CD79B\", \"IGHM\", \"JCHAIN\", \"TNFRSF13B\"], groupby=\"leiden_res_1.0\", standard_scale = \"var\", color_map = \"Blues\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21d7b861",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cluster 11 looks like B cell\n",
    "sc.tl.rank_genes_groups(raw_adata_cd4, groupby=\"leiden_res_1.0\", groups=['11'], method=\"wilcoxon\")\n",
    "sc.pl.rank_genes_groups(raw_adata_cd4, groupby=\"leiden_res_1.0\", n_genes=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00197198",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(sc.get.rank_genes_groups_df(raw_adata_cd4, group='11').head(30)[\"names\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63ceda84",
   "metadata": {},
   "outputs": [],
   "source": [
    "# check core marker expression overlays for cluster 11\n",
    "cluster_11_cells = raw_adata_cd4[raw_adata_cd4.obs['leiden_res_1.0'] == '11']\n",
    "sc.pl.umap(\n",
    "    cluster_11_cells,\n",
    "    color=[\"leiden_res_1.0\",\"CD4\",\"CD3D\",\"IL7R\",     # CD4 T markers\n",
    "           'CD19',\"MS4A1\",\"CD79A\",\"CD79B\",\"CD74\", # B cell markers\n",
    "           \"NKG7\",\"GNLY\",\"GZMB\"],   # NK / cytotoxic markers\n",
    "    vmax=\"p99\",\n",
    "    cmap=\"Reds\",\n",
    "    size=30,\n",
    "    save=\"_cluster11_marker_overlays.png\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5715bdf8",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.violin(raw_adata_cd4, ['n_genes_by_counts', 'total_counts', 'pct_counts_mt'], groupby= 'leiden_res_1.0', jitter=0.4, multi_panel=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eab02470",
   "metadata": {},
   "outputs": [],
   "source": [
    "raw_adata_cd4.obs['cell_type_cd4_res_1'] = raw_adata_cd4.obs['cell_type'].copy()\n",
    "adata_cd4_k11 = raw_adata_cd4.copy()\n",
    "sc.tl.leiden(adata_cd4_k11, resolution=1.0, restrict_to=('leiden_res_1.0', ['11']), key_added='leiden_res_1.0_k11')\n",
    "sc.pl.umap(adata_cd4_k11, color=['leiden_res_1.0', 'leiden_res_1.0_k11','cell_type_cd4_res_1'], legend_loc=\"on data\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e636eb21",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.dotplot(adata_cd4_k11, present_marker_genes , groupby=\"leiden_res_1.0_k11\", standard_scale=\"var\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0c8ba9c",
   "metadata": {},
   "source": [
    "Cluster 11 exhibited co-expression of canonical B-cell markers (MS4A1, CD79A) and pan-T-cell markers (CD3D, CD3E), suggesting the presence of T/B doublets. This hybrid signature, together with its distinct position on the UMAP, indicates technical doublets rather than a true biological subset. Consequently, this cluster was excluded from downstream analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b727749c",
   "metadata": {},
   "outputs": [],
   "source": [
    "raw_adata_cd4.obs['leiden_res_1.0'] = raw_adata_cd4.obs['leiden_res_1.0'].astype(str)\n",
    "raw_adata_cd4 = raw_adata_cd4[raw_adata_cd4.obs['leiden_res_1.0'] != '11', :].copy()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e20b370",
   "metadata": {},
   "outputs": [],
   "source": [
    "raw_adata_cd4.obs['cell_type_cd4_res_1'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf77fe9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(\n",
    "    raw_adata_cd4,\n",
    "    color=\"pica_id\",\n",
    "    size=2,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99d5bf00",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save data without cluster 11\n",
    "raw_adata_cd4.write_h5ad(\"/Users/jessica/Documents/GitHub/Jessica_notebooks/write/02_batch_annotation/PICA0001-PICA0007/PICA0001-PICA0007_preprocessed.h5ad\", compression='gzip')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "649d42d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_cd4= sc.read_h5ad(\"/Users/jessica/Documents/GitHub/Jessica_notebooks/write/02_batch_annotation/PICA0001-PICA0007/PICA0001-PICA0007_preprocessed.h5ad\")\n",
    "adata_cd4.var_names_make_unique()\n",
    "print(adata_cd4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05a9115d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# run scvi again\n",
    "scvi.model.SCVI.setup_anndata(adata_cd4, layer=\"counts\", batch_key=\"pica_id\")\n",
    "model = scvi.model.SCVI(adata_cd4, n_layers=2, n_latent=30, gene_likelihood=\"nb\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4cce0694",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.train()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f13884a",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_cd4.obsm[\"X_scVI\"] = model.get_latent_representation()\n",
    "\n",
    "#  latent representation \n",
    "sc.pp.neighbors(adata_cd4, use_rep=\"X_scVI\")\n",
    "sc.tl.umap(adata_cd4)\n",
    "\n",
    "\n",
    "sc.pl.umap(adata_cd4, color=[ \"pica_id\", 'cell_type'], save=\"_scvi_integrated_umap.png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "deaf6aae",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(\n",
    "    adata_cd4,\n",
    "    color=[\"log1p_total_counts\", \"pct_counts_mt\", \"log1p_n_genes_by_counts\"],\n",
    "    wspace=0.5,\n",
    "    ncols=2,\n",
    "    save=\"_scvi_qc.png\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7da3ef0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the scVI integrated batch adata\n",
    "adata_cd4.write_h5ad(\"/Users/jessica/Documents/GitHub/Jessica_notebooks/write/02_batch_annotation//PICA0001-PICA0007/PICA0001-PICA0007_scVI.h5ad\", compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b758083",
   "metadata": {},
   "outputs": [],
   "source": [
    "# clustering\n",
    "for res in [ 0.5, 1.0, 1.5, 2.0]:\n",
    "    sc.tl.leiden(\n",
    "        adata_cd4, key_added=f\"leiden_res_{res:3.1f}\", resolution=res, flavor=\"igraph\"\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e96e4ef0",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(\n",
    "    adata_cd4,\n",
    "    color=[ \"leiden_res_0.5\", \"leiden_res_1.0\",\"leiden_res_1.5\", \"leiden_res_2.0\"],\n",
    "    legend_loc=\"on data\", legend_fontsize = 15\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14e8b34b",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_cd4.write_h5ad(\"/Users/jessica/Documents/GitHub/Jessica_notebooks/write/02_batch_annotation/PICA0001-PICA0007/PICA0001-PICA0007_scVI_cluster.h5ad\", compression='gzip')\n",
    "print(adata_cd4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "291a144b",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.dotplot(adata_cd4, present_marker_genes , groupby=\"leiden_res_1.0\", standard_scale=\"var\", save=\"_all_markers.png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d38ab86",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_cd4.write_h5ad(\"/Users/jessica/Documents/GitHub/Jessica_notebooks/write/02_batch_annotation//PICA0001-PICA0007/PICA0001-PICA0007_cd4_annotations.h5ad\", compression='gzip')\n",
    "print(adata_cd4)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scanpy",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
